q2_demux/_demux.py

```diff
@@ -1,54 +1,123 @@
         self.barcode_generator = barcode_generator
         self.sequence_generator = sequence_generator
 
     def __iter__(self):
         # Adapted from q2-types
-        for barcode, sequence in itertools.zip_longest(
+        for barcode_fields, sequence_fields in itertools.zip_longest(
                 self.barcode_generator, self.sequence_generator):
-            if barcode is None:
+            if barcode_fields is None:
                 raise ValueError('More sequences were provided than barcodes.')
-            if sequence is None:
+            if sequence_fields is None:
                 raise ValueError('More barcodes were provided than sequences.')
             # The id or description fields may end with "/read-number", which
-            # will differ between the sequence and barcode reads. Confirm that
-            # they are identical up until the last /
-            if _extract_common_header(barcode.metadata['id']) != \
-               _extract_common_header(sequence.metadata['id']):
+            # and read number will differ between the sequence and barcode
+            # reads. Confirm that they are identical up until the last /
+            barcode_header_fields = \
+                barcode_fields[0][1:].split(' ', maxsplit=1)
+            sequence_header_fields = \
+                sequence_fields[0][1:].split(' ', maxsplit=1)
+
+            # confirm that both barcode and sequene have header lines with
+            # an equal number of space-separated fields
+            if len(barcode_header_fields) != len(sequence_header_fields):
+                raise ValueError('Mismatched header lines: %s and %s' %
+                                 (barcode_fields[0], sequence_fields[0]))
+
+            # confirm that the id fields are equal
+            if _extract_common_header(barcode_header_fields[0]) != \
+               _extract_common_header(sequence_header_fields[0]):
                 raise ValueError(
                     'Mismatched sequence ids: %s and %s' %
-                    (_extract_common_header(barcode.metadata['id']),
-                     _extract_common_header(sequence.metadata['id'])))
-            if _extract_common_header(barcode.metadata['description']) != \
-               _extract_common_header(sequence.metadata['description']):
-                raise ValueError(
-                    'Mismatched sequence descriptions: %s and %s' %
-                    (_extract_common_header(barcode.metadata['description']),
-                     _extract_common_header(sequence.metadata['description'])))
-            yield barcode, sequence
+                    (_extract_common_header(barcode_header_fields[0]),
+                     _extract_common_header(sequence_header_fields[0])))
+
+            # if a description field is present, confirm that they're equal
+            # note that the number of fields can only be 1 or 2, due to
+            # maxsplit
+            if len(barcode_header_fields) == 2:
```

Comment:
Can you also verify that the unit tests are testing these different cases? They might already be